modules/calib3d/src/calibration.cpp

```diff
@@ -1716,8 +1862,47 @@ CV_IMPL double cvCalibrateCamera2( const CvMat* objectPoints,
                     CvSize imageSize, CvMat* cameraMatrix, CvMat* distCoeffs,
                     CvMat* rvecs, CvMat* tvecs, int flags, CvTermCriteria termCrit )
 {
-    return cvCalibrateCamera2Internal(objectPoints, imagePoints, npoints, imageSize, cameraMatrix,
-                                      distCoeffs, rvecs, tvecs, NULL, NULL, flags, termCrit);
+    return cvCalibrateCamera2Internal(objectPoints, imagePoints, npoints, imageSize, -1, cameraMatrix,
+                                      distCoeffs, rvecs, tvecs, NULL, NULL, NULL, flags, termCrit);
+}
+
+CV_IMPL double cvCalibrateCamera4( const CvMat* objectPoints,
+                    const CvMat* imagePoints, const CvMat* npoints,
+                    CvSize imageSize, int iFixedPoint, CvMat* cameraMatrix, CvMat* distCoeffs,
+                    CvMat* rvecs, CvMat* tvecs, CvMat* newObjPoints, int flags, CvTermCriteria termCrit )
+{
+    // If iFixedPoint is out of rational range, fall back to standard method
+    if( iFixedPoint > npoints->data.i[0] - 2 )
+        iFixedPoint = -1;
+    // check object points. If not qualified, fall back to standard calibration.
+    int nimages = npoints->rows * npoints->cols;
+    int npstep = npoints->rows == 1 ? 1 : npoints->step / CV_ELEM_SIZE(npoints->type);
+    int i, ni;
+    if( iFixedPoint > 0 )
```

Comment:
Why `iFixedPoint == 0` is out of scope? Is it allowed?

> The index of the 3D object point in objectPoints[0] to be fixed. Usually it is
the top-right corner point of the calibration board grid.

Index is zero-based in C++. Zero index should correspond to "top-left" corner.

Reply:
Based on the cited paper, three 3D object points are chosen to be fixed. All x, y, z coordinates of point1 and point2 are fixed, and only z of point3 is fixed. In theory, these points could be arbitrarily chosen as long as they are non-collinear. but in practice, it will deliver more accurate results to select points that are far away from each other. In the internal implementation, the first (zero index, "top-left" corner) point is already selected as point1, see line 1651-1653. And the z coordinate of the last point ("bottom-right" corner) is also fixed as point3.z, see line 1657. Users have the freedom to select point2, and "top-right" or "bottom-left" point of the calibration board grid is recommended.

Use the convention in the sample C++ code, the first point ("top-left" corner) has the coordinates (0, 0, 0), "top-right" corner has (d, 0, 0) and the last point ("bottom-right" corner) has (x3, y3, 0). The accuracy of intrinsic parameters is free from these fixed points. But for accurate rvecs, tvecs and newObjPoints, the fixed coordinates must be accurate enough. In the above convention, only "d" should be accurately determined. "d" can be easily measured as the distance between the "top-left" and "top-right" corners.

There're samples on how to set the accurate "d". `samples/cpp/calibration.cpp` line 162 and `samples/cpp/tutorial_code/calib3d/camera_calibration/camera_calibration.cpp` line 561.

Reply:
Thank you for detailed explanation.

Perhaps we should have documentation notes about this (specify valid / expected range of input values):
> In the internal implementation, the first (zero index, "top-left" corner) point is already selected as point1

> And the z coordinate of the last point ("bottom-right" corner) is also fixed as point3.z

Also add parameters checks into code like this:
```
CV_Assert(iFixedPoint != 0);
CV_Assert(iFixedPoint != numpoints - 1);
```
(not sure about second one)

Reply:
The current document for `iFixedPoint` has the following information:

```
...
According to \cite strobl2011iccv, two other points are also fixed. In
this implementation, objectPoints[0].front and objectPoints[0].back.z
are used. Accurate rvecs and tvecs are only possible if coordinates of
these three fixed points are accurate enough.
...
```

I'd like to update it with the following text:

```
@param iFixedPoint The index of the 3D object point in objectPoints[0] to be fixed. Usually
the top-right corner point of the calibration board grid is recommended. For values out of the range
[1, objectPoints[0].size()-2], calibration will fall back to standard method. According to
\cite strobl2011iccv, two other points are also fixed. In this implementation, objectPoints[0].front
and objectPoints[0].back.z are used. Accurate rvecs, tvecs and newObjPoints are only possible if
coordinates of these three fixed points are accurate enough. If standard calibration method to be
used, just set this parameter to a negative value.
```

Is it acceptable?

------------------------------------------------------------------------

I think soft limit is a better experience for users, i.e. just falling back
to standard calibration method if `iFixedPoint < 1` or `iFixedPoint > numpoints - 2`.
If accuracy isn't the most important thing for some users, then fallback
scheme is acceptable for them. It seems assertions are a little bit
harsh for this parameter.

Reply:
Before using, users would like to compare accuracy between old and new functions. What parameters are fit for their tasks.

I believe we should not confuse them with the same results (and silently disabled algorithm) if `iFixedPoint` is out of valid range for this algorithm (error message is a normal way to say that input values is out of expected range - you may use `CV_Check()` macros to add more accurate error message).

Reply:
I also agree that silently falling back to the classical `calibrateCamera` is not a good thing.

Another [example](https://github.com/opencv/opencv/issues/5065) that led users to report issues are the `SOLVEPNP_DLS` and `SOLVEPNP_UPNP` flags in `SolvePnP` function.

Reply:
OK, the confusing might arise from the fact that we are using
`iFixedPoint` as a switch for calibration method selection. It'd be
better to not use "fall back" words. Just tell users to select new
method, pass in the parameter in the specified range. If standard method
is demanded, set it out of the range.

If new method is requested, but the `objectPoints` data are not qualified,
then `CV_Assert()` or `CV_error()` is called.

Thanks for your comments. I'll add more commits for further review.